### PR TITLE
Improve Makefile and fix systemd paths

### PR DESCRIPTION
**Pull request by 1dot75cm**
- add `DESTDIR` variable
- use` GOBUILD` instead of `go build` command
- use user's config files (`$HOME/.config/popub/*`)
@m13253

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,23 +3,24 @@
...
 clean:
 	rm -f popub-local/popub-local popub-relay/popub-relay
 
 install: all
-	install -Dm0755 popub-local/popub-local "$(PREFIX)/bin/popub-local"
-	install -Dm0755 popub-relay/popub-relay "$(PREFIX)/bin/popub-relay"
-	$(MAKE) -C systemd install
+	install -Dm0755 popub-local/popub-local "$(DESTDIR)$(PREFIX)/bin/popub-local"
+	install -Dm0755 popub-relay/popub-relay "$(DESTDIR)$(PREFIX)/bin/popub-relay"
+	$(MAKE) -C systemd install $(DESTDIR) $(PREFIX)
```
Comment (m13253) [inline Makefile]: Should this be `"DESTDIR=$(DESTDIR)" "PREFIX=$(PREFIX)"`?

File: systemd/Makefile [hunk 2]
```diff
@@ -1,13 +1,27 @@
...
+	install -Dm0644 example-local.conf $(DESTDIR)/etc/popub/local.d/example
+	install -Dm0644 example-relay.conf $(DESTDIR)/etc/popub/relay.d/example
 	systemctl daemon-reload || true
 
 uninstall:
-	rm -f /usr/lib/systemd/system/popub-local@.service /usr/lib/systemd/system/popub-relay@.service
-	rm -rf /etc/popub/
+	rm -f $(DESTDIR)$(SYSTEMD_UNIT_DIR)/popub-{local,relay}@.service
+	rm -f $(DESTDIR)$(SYSTEMD_USER_DIR)/popub-{local,relay}@.service
+	rm -rf $(DESTDIR)/etc/popub/
```
Comment (m13253) [inline systemd/Makefile]: Although most build services guarantee no space in `$DESTDIR`, would you please add quotes to ensure nothing will break if it contains space (or any other characters)?

Thanks.

**Review by m13253 — COMMENTED** [review]
Thank you for your improvements.
Here are some issues I think we might need to consider.

File: systemd/popub-local@.service.in [hunk 3]
```diff
@@ -3,14 +3,15 @@ Description=Popub Local
 After=network-online.target
 
 [Service]
+Type=simple
 AmbientCapabilities=CAP_NET_BIND_SERVICE
```
Comment (m13253) [inline systemd/popub-local@.service.in:7]: Not sure if we need to disable `CAP_NET_BIND_SERVICE` if it is a user service.

**Comment by 1dot75cm** [comment]
Thank you for your reply. I added the quotes to Makefile. @m13253

**Review by m13253 — COMMENTED** [review]
Thank you for your modification. Here is another problem I just found:

Not sure if we need to disable `CAP_NET_BIND_SERVICE` if it is a `systemd --user` service.

If it is run as a system service, `CAP_NET_BIND_SERVICE` is given because it "should" have the permission to bind to lower ports, but it is demoted to `nobody` thus not having the permission.

But it might be a security problem if `CAP_NET_BIND_SERVICE` is given to a non-root process, running as user service.

Since you wrote the user `.service` files, it's better to be careful.

Do you have any comments? Is it totally not a problem?

**Comment by 1dot75cm** [comment]
I agree with you. Of course, users need to decide if they want to bind the lower ports.
I fixed it by modify the Makefile. @m13253

**Comment by m13253** [comment]
On a shared user system, non-root user should not be able to bind to lower ports by default.
It is better to remove AmbientCapabilities for non-root user.
I am not sure if we need a user scope service. Is it worth doing this?

**Comment by 1dot75cm** [comment]
You can provide a user scope service and users decide to if they want to do. I have removed AmbientCapabilities for non-root user. @m13253

**Comment by m13253** [comment]
Have you pushed the updated code? For some reason I am still seeing old version.

**Comment by m13253** [comment]
Okay that was a cache problem.
Merging now.

**Comment by m13253** [comment]
Code causing problems if environment file in `.local` is missing (I'm starting a system service):
```
● popub-local@ssh.service - Popub Local
   Loaded: loaded (/usr/lib/systemd/system/popub-local@.service; disabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: resources) since Sun 2017-10-01 04:27:33 CST; 2s ago
 Main PID: 2590 (code=killed, signal=TERM)

Oct 01 04:27:33 brilliant-rpi systemd[1]: popub-local@ssh.service: Failed to load environment files: No such file or directory
Oct 01 04:27:33 brilliant-rpi systemd[1]: popub-local@ssh.service: Failed to run 'start' task: No such file or directory
Oct 01 04:27:33 brilliant-rpi systemd[1]: Failed to start Popub Local.
Oct 01 04:27:33 brilliant-rpi systemd[1]: popub-local@ssh.service: Unit entered failed state.
Oct 01 04:27:33 brilliant-rpi systemd[1]: popub-local@ssh.service: Failed with result 'resources'.
```
I can't bear if we need these too many ugly `sed`s to provide a "user" service where it is less secure and nobody would use, and completely unable to start at all.

**Comment by m13253** [comment]
I found you modified the path from `/etc/popub/local/example.conf` to `/etc/popub/local.d/example` without changing the documentation.
Is it intended behavior? Or is it required by Red Hat?

I was spending double time diagnosing this change without being noticed.

**Comment by m13253** [comment]
Have you been testing your unit files? The user service does not work either:
```
● popub-local@ssh.service - Popub Local
   Loaded: loaded (/usr/lib/systemd/user/popub-local@.service; disabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: exit-code) since Sun 2017-10-01 04:41:39 CST; 2s ago
  Process: 2742 ExecStart=/usr/local/bin/popub-local $LOCAL_ADDR $RELAY_ADDR $AUTH_KEY (code=exited, status=216/GROUP)
 Main PID: 2742 (code=exited, status=216/GROUP)

Oct 01 04:41:39 brilliant-rpi systemd[683]: popub-local@ssh.service: Unit entered failed state.
Oct 01 04:41:39 brilliant-rpi systemd[683]: popub-local@ssh.service: Failed with result 'exit-code'.
```

I am really disappointed that you did not even test your code before submitting. 😞

**Pull request merged**